src/Razor/src/Microsoft.CodeAnalysis.Razor.CohostingShared/HtmlDocumentServices/HtmlDocumentSynchronizer.cs

```diff
@@ -48,18 +48,19 @@ public async Task<SynchronizationResult> TrySynchronizeAsync(TextDocument docume
 
         _logger.LogDebug($"TrySynchronize for {document.FilePath} as at {requestedVersion}");
 
-        // We are not passing on the cancellation token through to the actual task that does the generation, because
-        // we do the actual work on whatever happens to be the first request that needs Html, without knowing how important
-        // that request is, nor why it might be cancelled. If that request is cancelled because of a document update,
-        // then the next request will cancel any work we start anyway, and if that request was cancelled for some other reason,
-        // then the next request probably wants the same version of the document, so we've got a head start.
-        return await GetSynchronizationRequestTaskAsync(document, requestedVersion).ConfigureAwait(false);
+        return await GetSynchronizationRequestTaskAsync(document, requestedVersion, cancellationToken).ConfigureAwait(false);
     }
 
-    private Task<SynchronizationResult> GetSynchronizationRequestTaskAsync(TextDocument document, RazorDocumentVersion requestedVersion)
+    private Task<SynchronizationResult> GetSynchronizationRequestTaskAsync(TextDocument document, RazorDocumentVersion requestedVersion, CancellationToken cancellationToken)
     {
         lock (_gate)
         {
+            if (cancellationToken.IsCancellationRequested)
```

Comment:
Why do we acquire a lock to cancel?

Reply:
I was more thinking "check if cancellation happened while we were waiting to acquire a lock".

Reply:
If there are several requests at once, none of them will cancel until they can acquire the lock. They'll all be stacked up behind it waiting to learn that they've been cancelled. Consider switching to a `SemaphoreSlim` or other lock that provides a cancellable `Wait`. Using a regular ol' `lock` will be a problem anyway, if this method every becomes `async`.

Reply:
Switched to semaphore slim